desktop/notification/fdo_test.go

```diff
@@ -0,0 +1,611 @@
+			return []*dbus.Message{response}, nil
+		default:
+			return nil, fmt.Errorf("unexpected message #%d: %s", n, msg)
+		}
+	})
+	err := srv.ObserveNotifications(context.TODO(), &testObserver{})
+	c.Assert(err, ErrorMatches, "org.freedesktop.DBus.Error.NameHasNoOwner")
+}
+
+func (s *fdoSuite) TestObserveNotificationsRemoveWatchError(c *C) {
+	logBuffer, restore := logger.MockLogger()
+	defer restore()
+
+	ctx, cancel := context.WithCancel(context.TODO())
+	msgsSeen := 0
+	addMatchSeen := make(chan struct{}, 1)
+	defer close(addMatchSeen)
+	srv := s.connectWithHandler(c, func(msg *dbus.Message, n int) ([]*dbus.Message, error) {
+		msgsSeen++
+		switch n {
+		case 0:
+			s.checkAddMatchRequest(c, msg)
+			response := s.addMatchResponse(c, msg)
+			addMatchSeen <- struct{}{}
+			return []*dbus.Message{response}, nil
+		case 1:
+			s.checkRemoveMatchRequest(c, msg)
+			response := s.nameHasNoOwnerResponse(c, msg)
+			return []*dbus.Message{response}, nil
+		default:
+			return nil, fmt.Errorf("unexpected message #%d: %s", n, msg)
+		}
+	})
+
+	var wg sync.WaitGroup
+	wg.Add(1)
+	go func() {
+		err := srv.ObserveNotifications(ctx, &testObserver{})
+		// The error from RemoveWatch is not clobbering the return value of ObserveNotifications.
+		c.Assert(err, ErrorMatches, "context canceled")
+		c.Check(logBuffer.String(), testutil.Contains, "Cannot remove D-Bus signal matcher: org.freedesktop.DBus.Error.NameHasNoOwner\n")
+		wg.Done()
+	}()
+	// Wait for the signal that we saw the AddMatch message and then stop.
+	<-addMatchSeen
+	cancel()
+	// Wait for ObserveNotifications to return
+	wg.Wait()
+	c.Check(msgsSeen, Equals, 2)
+}
+
+func (s *fdoSuite) TestObserveNotificationsProcessingError(c *C) {
+	msgsSeen := 0
+	srv := s.connectWithHandler(c, func(msg *dbus.Message, n int) ([]*dbus.Message, error) {
+		msgsSeen++
+		switch n {
+		case 0:
+			s.checkAddMatchRequest(c, msg)
+			response := s.addMatchResponse(c, msg)
+			sig := &dbus.Message{
+				Type: dbus.TypeSignal,
+				Headers: map[dbus.HeaderField]dbus.Variant{
+					dbus.FieldPath:      dbus.MakeVariant(dbus.ObjectPath("/org/freedesktop/Notifications")),
+					dbus.FieldInterface: dbus.MakeVariant("org.freedesktop.Notifications"),
+					dbus.FieldMember:    dbus.MakeVariant("ActionInvoked"),
+					dbus.FieldSender:    dbus.MakeVariant("org.freedesktop.Notifications"),
+					dbus.FieldSignature: dbus.MakeVariant(dbus.SignatureOf(uint32(0), "")),
+				},
+				Body: []interface{}{uint32(42), "action-key"},
+			}
+			// Send the DBus response for the method call and an additional signal.
+			return []*dbus.Message{response, sig}, nil
+		case 1:
+			s.checkRemoveMatchRequest(c, msg)
+			response := s.removeMatchResponse(c, msg)
+			return []*dbus.Message{response}, nil
+		default:
+			return nil, fmt.Errorf("unexpected message #%d: %s", n, msg)
+		}
+	})
+	err := srv.ObserveNotifications(context.TODO(), &testObserver{
+		actionInvoked: func(id notification.ID, actionKey string) error {
```

Comment:
(nitpick) this (and similar ones) might increase a counter to ensure the action was only invoked once.